fem/pgridfunc.cpp

```diff
@@ -249,6 +249,8 @@ void ParGridFunction::ExchangeFaceNbrData()
    auto send_data_ptr = mpi_gpu_aware ? send_data.Read() : send_data.HostRead();
    auto face_nbr_data_ptr = mpi_gpu_aware ? face_nbr_data.Write() :
                             face_nbr_data.HostWrite();
+   // Wait for the kernel to be done since it updates what's sent and it may be async
+   if (mpi_gpu_aware) { MFEM_DEVICE_SYNC; }
```

Comment:
Isn't MPI supposed to sync to the default stream 0 that we use? I think all streams sync to that.

Reply:
I should have put this in the description, the reason I added it was that we have tests that fail without the sync. I asked a few people and they confirmed that we _should_ need it, so that's consistent

Reply:
I want to understand why it is failing in your test. Maybe it is specific to HIP -- the stream synchronizations may be a bit different from CUDA. Let's chat offline.

Reply:
This came up in the meeting today, but this comment suggests that @tomstitt's proposed fix is correct for CUDA-aware Open MPI: https://github.com/open-mpi/ompi/issues/7733#issuecomment-648343243.

Reply:
^ +1, since MFEM puts all kernels on stream 0 (as far as I know), we should be fine to just use `MFEM_STREAM_SYNC` instead of the generally slower `MFEM_DEVICE_SYNC`.

Reply:
Maybe a corner case is when RAJA is configured to run on a stream different from 0. Even then, the sync to stream 0 may actually sync to other streams as well, I'm not sure.

Reply:
> Maybe a corner case is when RAJA is configured to run on a stream different from 0. Even then, the sync to stream 0 may actually sync to other streams as well, I'm not sure.

Yes! I was typing out my response but you summed it up

Reply:
The CUDA documentation about `cudaStreamSynchronize`, e.g. [here](https://docs.nvidia.com/cuda/archive/11.2.2/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g82b5784f674c17c6df64affe618bf45e), says that

  _"This function uses standard [default stream](https://docs.nvidia.com/cuda/archive/11.2.2/cuda-runtime-api/stream-sync-behavior.html#stream-sync-behavior__default-stream) semantics."_

And the documentation pointed to by the link says:

_"The legacy default stream is an implicit stream which synchronizes with all other streams in the same `CUcontext` except for non-blocking streams, described below."_

To me this suggests that synchronizing with stream 0 also synchronizes with other streams except for non-blocking streams. I suspect that the default RAJA/CAMP stream is not non-blocking, so using `MFEM_STREAM_SYNC` should be okay.

The HIP documentation about `hipStreamSynchronize`, e.g. [here](https://rocm.docs.amd.com/projects/HIP/en/latest/doxygen/html/group___stream.html#gabbfb9f573a6ebe8c478605ecb5504a74) says:

_"This command follows standard null-stream semantics. Specifically, specifying the null stream will cause the command to wait for other streams on the same device to complete all pending operations."_

so we should be okay in HIP context as well.

Reply:
thanks @v-dobrev !